crush/crush.go

```diff
@@ -101,6 +103,38 @@ func CreateTarGz(destination io.Writer, source string) error {
 	return CreateTar(gz, source)
 }
 
+// ExtractArchive extracts source archive to a destination directory.  An arbitrary number of top-level directory
+// components can be stripped from each path.
+func ExtractArchive(source *os.File, destination string, stripComponents int) error {
+	b, err := ioutil.ReadFile(source.Name())
+	if err != nil {
+		return err
+	}
+	kind, err := filetype.Match(b)
+	if err != nil {
+		return err
+	}
+	
+	if kind == filetype.Unknown {
+		return fmt.Errorf("unknown file type")
+	}
+
+	switch kind.MIME.Value {
+	case "application/x-tar":
+		return ExtractTar(source, destination, stripComponents)
+	case "application/x-bzip2":
+		return ExtractTarBz2(source, destination, stripComponents)
+	case "application/gzip":
+		return ExtractTarGz(source, destination, stripComponents)
```

Comment:
This assumes that you have a `.tar.gz` or `.tar.bz2` file. You can in fact just have a `.gz` file, like if you gzip a go binary executable.

I think this needs to be handled differently. My thought would be to have an internal `extractArchive(source io.Reader, destination string, stripComponents int)` method. I think you could then do something like this:

```
case "application/x-bzip2":
    return extractArchive(bzip2.NewReader(source), destination, stripComponents)
...
```

Basically, pass in a reader that unzip's it and look at what's in that file. I think you can use [MatchReader](https://github.com/h2non/filetype/blob/master/match.go#L55) instead of Match and it should be able to read from the reader and match against that.